### PR TITLE
Automatically check python code in documentation

### PR DESCRIPTION
**Pull request by jigold**
- added `doctest` step to `makeHailDocs` in gradle script
 - All example files are in `python/hail/docs/data/`
 - All output files in code should have the path `output/`
 - For multi-line code statements, only first line can have `>>>`; all other lines must start with `...`
 - Modified `doctest` to **NOT** check output. This is different than how `doctest` normally works.
 - added overwrite option to python HailContext and the `write` command in scala.
 - added more classes to Hail namespace (`from hail import *`)

 For `data/example.vds`, the code used to generate it is:

 ```
 hc.import("src/test/resources/sample.vcf.bgz").downsample_variants(10).annotate_variants_expr('va.useInKinship = pcoin(0.9), va.panel_maf = 0.1, va.anno1 = 5, va.anno2 = 0, va.consequence = "LOF", va.gene = "A", va.score = 5.0').split_multi().variant_qc().sample_qc().annotate_samples_expr('sa.isCase = true, sa.pheno.isCase = pcoin(0.5), sa.pheno.isFemale = pcoin(0.5), sa.pheno.age=rnorm(65, 10), sa.cov.PC1 = rnorm(0,1), sa.pheno.height = rnorm(70, 10), sa.cov1 = rnorm(0, 1), sa.cov2 = rnorm(0,1), sa.pheno.bloodPressure= rnorm(120,20), sa.pheno.cohortName = "cohort1"').write("python/hail/docs/data/example.vds", overwrite=True)
 ```

 For `data/exampe2.vds`, the code is:

 ```
 hc.import("src/test/resources/sample.vcf.bgz").downsample_variants(5).annotate_variants_expr('va.anno1 = 5, va.toKeep1 = true, va.toKeep2 = false, va.toKeep3 = true').split_multi().write("python/hail/docs/data/example2.vds", overwrite=True)
 ```

 For `data/example_lmmreg.vds`, the code is:

 ```
 hc.import_vcf('src/test/resources/sample.vcf')
   .split_multi()
   .variant_qc()
   .annotate_samples_expr('sa.culprit = gs.filter(g => v == Variant("20", 13753124, "A", "C")).map(g => g.gt).collect()[0]')
   .annotate_samples_expr('sa.pheno = rnorm(1,1) * sa.culprit')
   .annotate_samples_expr('sa.cov1 = rnorm(0,1)')
   .annotate_samples_expr('sa.cov2 = rnorm(0,1)')
   .linreg('sa.pheno', ['sa.cov1', 'sa.cov2']).annotate_variants_expr('va.useInKinship = va.qc.AF > 0.05')
   .write("python/hail/docs/data/example_lmmreg.vds", overwrite=True)
 ```

### CODE REVIEW

**Comment by jbloom22** [comment]
The linreg example above uses linreg, not lmmreg.

**Comment by jbloom22** [comment]
This is awesome. There are a couple more command in Getting Started you could add.

**Comment by jigold** [comment]
@jbloom22 -- added the commands in getting started

File: python/hail/docs/conf.py [hunk 1]
```diff
@@ -45,6 +49,30 @@
...
+    if os.path.isdir(vds):
+        shutil.rmtree(vds)
+
+hc = HailContext(log="output/hail.log", quiet=True)"""
+
+doctest_global_cleanup = """import shutil, os
+
+hc.stop()
+
+shutil.rmtree("output/")
```
Comment (tpoterba) [inline python/hail/docs/conf.py]: would prefer cleanup at beginning, so failed tests don't cause subsequent builds to fail.  Or both

File: python/hail/keytable.py [hunk 2]
```diff
@@ -9,6 +9,40 @@
...
 
+    KeyTables can be imported from a text file with :py:meth:`~hail.HailContext.import_keytable` or generated from an existing
+    VariantDataset with :py:meth:`~hail.VariantDataset.aggregate_by_key`, :py:meth:`~hail.VariantDataset.make_keytable`,
+    :py:meth:`~hail.VariantDataset.samples_keytable`, and :py:meth:`~hail.VariantDataset.variants_keytable`.
+
+    In the examples below, we have imported two KeyTables from text files (``kt1`` and ``kt2``).
+
+    >>> kt1 = hc.import_keytable("data/kt_example1.tsv", ["ID"], config=TextTableConfig(impute=True))
+
+    +--+---+---+-+-+----+----+----+
```
Comment (tpoterba) [inline python/hail/keytable.py]: Not a comment on this PR, but we should make a show() method on keytables like dataframes.  or just make it easy to do that with a dataframe.

File: python/hail/keytable.py [hunk 3]
```diff
@@ -89,10 +146,8 @@ def same(self, other):
 
         **Examples**
 
-        >>> kt1 = hc.import_keytable("data/example1.tsv")
-        >>> kt2 = hc.import_keytable("data/example2.tsv")
         >>> if kt1.same(kt2):
-        >>>     print_function("KeyTables are the same!")
+        ...     print_function("KeyTables are the same!")
```
Comment (tpoterba) [inline python/hail/keytable.py]: why print_function?

File: python/hail/keytable.py [hunk 4]
```diff
@@ -303,9 +338,8 @@ def forall(self, code):
 
         Test whether all rows in the KeyTable have the value of ``C1`` equal to 5:
 
-        >>> kt = hc.import_keytable('data/example.tsv')
-        >>> if kt.forall("C1 == 5"):
-        >>>     print_function("All rows have C1 equal 5.")
+        >>> if kt1.forall("C1 == 5"):
+        ...     print_function("All rows have C1 equal 5.")
```
Comment (tpoterba) [inline python/hail/keytable.py]: again, should be just `print(...)`

File: python/hail/keytable.py [hunk 5]
```diff
@@ -324,9 +358,8 @@ def exists(self, code):
 
         Test whether any row in the KeyTable has the value of ``C1`` equal to 5:
 
-        >>> kt = hc.import_keytable('data/example.tsv')
-        >>> if kt.exists("C1 == 5"):
-        >>>     print_function("At least one row has C1 equal 5.")
+        >>> if kt1.exists("C1 == 5"):
+        ...     print_function("At least one row has C1 equal 5.")
```
Comment (tpoterba) [inline python/hail/keytable.py]: again print

File: src/main/scala/is/hail/driver/ExportVCF.scala [hunk 6]
```diff
@@ -23,6 +23,8 @@ object ExportVCF extends Command {
 
     @Args4jOption(name = "--parallel", usage = "Export VCF in parallel")
     var parallel: Boolean = false
+
```
Comment (tpoterba) [inline src/main/scala/is/hail/driver/ExportVCF.scala]: whitespace change (causes merge conflicts elsewhere)

File: src/main/scala/is/hail/driver/Main.scala [hunk 7]
```diff
@@ -15,7 +15,6 @@ import scala.reflect.ClassTag
 
 object HailConfiguration {
   var tmpDir: String = "/tmp"
-
```
Comment (tpoterba) [inline src/main/scala/is/hail/driver/Main.scala]: whitespace chnage

File: python/hail/representation/variant.py [hunk 8]
```diff
@@ -145,15 +145,17 @@ def num_alleles(self):
...
 
-         The reference is included in the allele index.  The index of
-         the first alternate allele is 1.  The following is true for all
-         variants:
+        The reference is included in the allele index.  The index of
+        the first alternate allele is 1.  The following is true for all
+        variants:
 
-         >>> v.ref == v.allele(0)
+        >>> v = Variant.parse("16:2478434:A:T,TA")
```
Comment (tpoterba) [inline python/hail/representation/variant.py]: I think this point no longer makes as much sense when you have to construct a variant here.  Maybe move the variant constructors above?

Comment (tpoterba) [inline python/hail/representation/variant.py]: (test setup like for the interval stuff)

File: python/hail/representation/interval.py [hunk 9]
```diff
@@ -99,7 +104,11 @@ def overlaps(self, interval):
 
         is equivalent to
 
-        >>> interval1.contains(interval2.start) or interval2.contains(interval1.start)
+        >>> interval1.contains(interval2.start)
+
+        or
```
Comment (tpoterba) [inline python/hail/representation/interval.py]: Breaking this out is incorrect, I think.  The statement should stay together.

File: python/hail/representation/genotype.py [hunk 10]
```diff
@@ -279,7 +293,7 @@ def fraction_reads_ref(self):
 
         Equivalent to:
 
-        >>> g.ad()[0] / sum(g.ad())
+        >>> g.ad[0] / sum(g.ad)
```
Comment (tpoterba) [inline python/hail/representation/genotype.py]: good catch!

**Comment by jigold** [comment]
@tpoterba Back to you.

File: python/hail/dataset.py [hunk 11]
```diff
@@ -14,6 +14,15 @@
 class VariantDataset(object):
     """Hail's primary representation of genomic data, a matrix keyed by sample and variant.
 
+    Variant Datasets can be generated from a VCF file with :py:meth:`~hail.HailContext.import_vcf`,
```
Comment (tpoterba) [inline python/hail/dataset.py]: I think this capitalization looks bad.  Either use code:

A `VariantDataset` can be generated from...

However, I think we can tighten all this up.  Basically everything on HailContext creates a VDS, so we should just link there.

File: python/hail/dataset.py [hunk 12]
```diff
@@ -461,17 +462,10 @@ def annotate_samples_fam(self, input, quantpheno=False, delimiter='\\\\s+', root
...
 
-        >>> vds.annotate_samples_fam("data/myStudy.fam", quantPheno=True)
-
-        Import case-control phenotype data from an
-        arbitrary-whitespace-delimited `PLINK .fam
-        <https://www.cog-genomics.org/plink2/formats#fam>`_ file into sample
-        annotations:
-
-        >>> vds.annotate_samples_fam("data/myStudy.fam", delimiter="\\s+")
+        >>> vds.annotate_samples_fam("data/myStudy.fam", quantpheno=True)
```
Comment (tpoterba) [inline python/hail/dataset.py]: add a binding here on the left hand side: `vds = vds.anno...`

File: python/hail/dataset.py [hunk 13]
```diff
@@ -406,22 +411,18 @@ def annotate_samples_expr(self, condition):
...
+        >>> (vds.annotate_variants_table('data/consequence.tsv', 'Variant', code='va.consequence = table.Consequence', config=TextTableConfig(impute=True))
+        ...     .annotate_variants_expr('va.isSingleton = gs.map(g => g.nNonRefAlleles).sum() == 1')
+        ...     .annotate_samples_expr('sa.LOF_genes = gs.filter(g => va.isSingleton && g.isHet && va.consequence == "LOF").map(g => va.gene).collect()'))
 
         To create an annotation for only a subset of samples based on an existing annotation:
 
-        >>> (hc.read('data/example.vds')
-        >>>    .annotate_samples_table('data/phenotypes.tsv', 'Sample', root='sa.pheno', config=TextTableConfig(impute=True))
-        >>>    .annotate_samples_expr('if (sa.pheno.cohortName == "cohort1") sa.pheno.bloodPressure else NA: Double')
+        >>> vds.annotate_samples_expr('sa.newpheno = if (sa.pheno.cohortName == "cohort1") sa.pheno.bloodPressure else NA: Double')
```
Comment (tpoterba) [inline python/hail/dataset.py]: Add a binding here on the left.

Comment (tpoterba) [inline python/hail/dataset.py]: Actually, you need to do this basically everywhere.

File: python/hail/dataset.py [hunk 14]
```diff
@@ -2261,9 +2231,9 @@ def lmmreg(self, kinship_vds, y, covariates=[], global_root="global.lmmreg", va_
 
         **Examples**
 
-        Suppose ``assoc.vds`` has a Boolean variant annotation ``va.useInKinship`` and numeric or Boolean sample annotations ``sa.pheno``, ``sa.cov1``, ``sa.cov2``. Then the :py:meth:`.lmmreg` function in
+        Suppose ``vds`` has a Boolean variant annotation ``va.useInKinship`` and numeric or Boolean sample annotations ``sa.pheno``, ``sa.cov1``, ``sa.cov2``. Then the :py:meth:`.lmmreg` function in
```
Comment (tpoterba) [inline python/hail/dataset.py]: incorrect statement about `vds`: you're reading example_lmmreg.vds, so say something like "the VDS saved at foo/bar/..."

File: python/hail/docs/getting_started.rst [hunk 15]
```diff
@@ -63,42 +67,58 @@ Add the following environmental variables and make an alias for Hail by filling
...
 
-To :func:`count <hail.VariantDataset.count>` the number of samples, variants, and genotypes, run::
+To :func:`count <hail.VariantDataset.count>` the number of samples, variants, and genotypes, run:
+
+.. doctest::
 
     >>> vds.count(genotypes=True)
 
-Now let's get a feel for Hail's powerful :ref:`objects <sec-objects>`, `annotation system <../reference.html#Annotations>`_, and `expression language <../reference.html#HailExpressionLanguage>`_. To :func:`print <hail.VariantDataset.print_schema>` the current annotation schema and use these annotations to filter variants, samples, and genotypes, run::
+Now let's get a feel for Hail's powerful :ref:`objects <sec-objects>`, `annotation system <../reference.html#Annotations>`_, and `expression language <../reference.html#HailExpressionLanguage>`_. To :func:`print <hail.VariantDataset.print_schema>` the current annotation schema and use these annotations to filter variants, samples, and genotypes, run:
```
Comment (tpoterba) [inline python/hail/docs/getting_started.rst]: print_schema no longer exists

File: python/hail/keytable.py [hunk 16]
```diff
@@ -9,6 +9,40 @@
 class KeyTable(object):
     """Hail's version of a SQL table where columns can be designated as keys.
 
+    KeyTables can be imported from a text file with :py:meth:`~hail.HailContext.import_keytable` or generated from an existing
```
Comment (tpoterba) [inline python/hail/keytable.py]: Don't capitalize like this in plain text -- just say "Key tables"

File: python/hail/keytable.py [hunk 17]
```diff
@@ -9,6 +9,40 @@
...
+    VariantDataset with :py:meth:`~hail.VariantDataset.aggregate_by_key`, :py:meth:`~hail.VariantDataset.make_keytable`,
+    :py:meth:`~hail.VariantDataset.samples_keytable`, and :py:meth:`~hail.VariantDataset.variants_keytable`.
+
+    In the examples below, we have imported two KeyTables from text files (``kt1`` and ``kt2``).
```
Comment (tpoterba) [inline python/hail/keytable.py]: same

**Comment by jigold** [comment]
@tpoterba Back to you

File: python/hail/dataset.py [hunk 18]
```diff
@@ -1789,17 +1763,17 @@ def filter_samples_expr(self, condition, keep=True):
...
 
-        >>> vds.filter_samples_expr('"^NA" ~ s' , keep=False)
+        >>> vds_result = vds.filter_samples_expr('"^NA" ~ s.id' , keep=False)
 
         Filter samples from sample QC metrics and write output to a new dataset:
 
-        >>> (vds.sample_qc()
-        >>>     .filter_samples_expr('sa.qc.callRate >= 0.99 && sa.qc.dpMean >= 10')
-        >>>     .write("data/filter_samples.vds"))
+        >>> vds_result = (vds.sample_qc()
```
Comment (tpoterba) [inline python/hail/dataset.py]: no result after a write

File: python/hail/dataset.py [hunk 19]
```diff
@@ -2040,13 +2010,11 @@ def ibd(self, output, maf=None, unbounded=False, min=None, max=None):
 
         To estimate and write the full IBD matrix to *ibd.tsv*, estimated using minor allele frequencies computed from the dataset itself:
 
-        >>> (hc.read('data/example.vds')
-        >>>  .ibd('data/ibd.tsv'))
+        >>> vds_result = vds.ibd('output/ibd.tsv')
```
Comment (tpoterba) [inline python/hail/dataset.py]: IBD doesn't return a vds, fix this and the one below

**Pull request merged**